Modules/Axis/Sources/Axis/String/UnicodeScalars.swift

```diff
@@ -0,0 +1,83 @@
+        "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z", "_",
+        "a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r",
+        "s", "t", "u", "v", "w", "x", "y", "z", "~"
+    ]
+
+    public static let uriPathAllowed: UnicodeScalars = [
+        "!", "$", "&", "\'", "(", ")", "*", "+", ",", "-", ".", "/", "0", "1", "2", "3", "4", "5",
+        "6", "7", "8", "9", ":", "=", "@", "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K",
+        "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z", "_", "a", "b",
+        "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t",
+        "u", "v", "w", "x", "y", "z", "~"
+    ]
+
+    public static let uriHostAllowed: UnicodeScalars = [
+        "!", "$", "&", "\'", "(", ")", "*", "+", ",", "-", ".", "0", "1", "2", "3", "4", "5", "6",
+        "7", "8", "9", ":", ";", "=", "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L",
+        "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z", "[", "]", "_", "a",
+        "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s",
+        "t", "u", "v", "w", "x", "y", "z", "~"
+    ]
+
+    public static let uriPasswordAllowed: UnicodeScalars = [
+        "!", "$", "&", "\'", "(", ")", "*", "+", ",", "-", ".", "0", "1", "2", "3", "4", "5", "6",
+        "7", "8", "9", ";", "=", "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M",
+        "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z", "_", "a", "b", "c", "d",
+        "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v",
+        "w", "x", "y", "z", "~"
+    ]
+
+    public static let uriUserAllowed: UnicodeScalars = [
+        "!", "$", "&", "\'", "(", ")", "*", "+", ",", "-", ".", "0", "1", "2", "3", "4", "5", "6",
+        "7", "8", "9", ";", "=", "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M",
+        "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z", "_", "a", "b", "c", "d",
+        "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v",
+        "w", "x", "y", "z", "~"
+    ]
+
+    private let scalars: Set<UnicodeScalar>
+
+    public init(scalars: Set<UnicodeScalar>) {
+        self.scalars = scalars
+    }
+
+    public init(arrayLiteral elements: UnicodeScalar...) {
+        self.init(scalars: Set(elements))
+    }
+
+    public func contains(_ scalar: UnicodeScalar) -> Bool {
+        return scalars.contains(scalar)
+    }
+
+    public func utf8() -> Set<UTF8.CodeUnit> {
+        var codeUnits: Set<UTF8.CodeUnit> = []
+        scalars.forEach {
+            UTF8.encode($0) {
+                codeUnits.insert($0)
+            }
+        }
+        return codeUnits
+    }
+}
+
+//extension UnicodeScalars {
```

Comment:
Sure.  I put it there to just keep it in line with other parts of the code.  Actually, I wanted to discuss that `Error`s are better readable and managed if they are nested in their parent types.  E.g. 

````swift
extension String {
    public enum Error: Swift.Error {
        case invalidUTF8
        case invalidPercentEncoding
    }
}
// ...
guard let hexH = reader.next(), let hexL = reader.next() else { throw String.Error.invalidPercentEncoding }
````

What do you think?  I can submit another PR.